google/cloud/bigquery/ipython_magics/line_arg_parser/lexer.py

```diff
@@ -0,0 +1,250 @@
+# Copyright 2020 Google LLC
```

Comment:
Have you considered using `pyparsing` for this? It appears to be a pretty acceptable dependency, and in past projects I've found it more intuitive for complex parsing than regular expressions.

If we don't move away from regex, I think I'd need a few paragraphs of explanation in comments to understand what's intended in the `token_types` state machine

Reply:
To be honest, I have not. Partially because I didn't want to introduce a new dependency "just" for this task, and partially because I already had a working prototype after two days, thus the task didn't feel all that complex.

In hindsight, it could indeed be a viable alternative, although I'm not sure if it's still worth making a switch at this point...

---

Re: maintenance - the goal of the states isolate the tokens from other tokens that can match similar patterns.

For instance, when parsing the `--params` option, we only want to be matching Python tokens, and not other stuff like OPT_VAL, which is very general and matches almost anything. On the same note, while parsing a non-`--params` option value, we don't need (or want) to be recognizing Python tokens and end up with a syntax error where no Python constructs are expected.

Pattern precedence is important when using regexes like that, and because some patterns overlap, it is not clear which of them should be listed (and tried) first. Lexer states thus serve as namespaces that isolate tokens groups and prevent the tokens from stepping on other unrelated  tokens' toes.

Does this explanation make sense? If it does, I can reword and and add it as a top-level comment in the lexer module.

Reply:
Thanks for the explanation. This helps.

I think I'd want a little more detail about why the order is the way it is. That way if we need to make another more specific parser for an argument in the future, we have an idea of where to slot it in.